src/ModelContextProtocol/Client/McpClientTool.cs

```diff
@@ -9,21 +9,50 @@ namespace ModelContextProtocol.Client;
 public sealed class McpClientTool : AIFunction
 {
     private readonly IMcpClient _client;
+    private readonly string? _name;
+    private readonly string? _description;
 
-    internal McpClientTool(IMcpClient client, Tool tool)
+    internal McpClientTool(IMcpClient client, Tool tool, string? name = null, string? description = null)
     {
         _client = client;
         ProtocolTool = tool;
+        _name = name ?? tool.Name;
+        _description = description ?? tool.Description;
+    }
+
+    /// <summary>
+    /// Creates a new instance of the tool with the specified name.
+    /// This is useful for optimizing the tool name for specific models or for prefixing the tool name with a (usually server-derived) namespace to avoid conflicts.
+    /// The server will still be called with the original tool name, so no mapping is required.
+    /// </summary>
+    /// <param name="name">The model-facing name to give the tool. Pass null to clear the name override and to use the MCP Tool name again.</param>
+    /// <returns>Copy of this McpClientTool with the provided name</returns>
+    public McpClientTool WithName(string? name)
```

Comment:
I know I approved this, by why allow null names here? Shouldn't it just throw in that case?

Reply:
I agree this should prohibit null. I'm not seeing a use case for wanting to revert back to the original name, and it doesn't actually let you revert back to the previous name, just the original.